manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+let g = ref None
+
+let d1 () =
+  let c = ref 0 in
+  c := 42;
+  let a = !c in
+  g := Some c;
+  a
+
+let d2 () =
+  match !g with
+  | None -> ()
+  | Some c -> c := 7
+\end{caml_example*}
+
+Observe that there is a data race on both "g" and "c". Consider only the first
+three instructions in "d1":
+
+\begin{verbatim}
+let c = ref 0 in
+c := 42;
+let a = !c in
+...
+\end{verbatim}
+
+The OCaml memory model is bound both in space and time. The only memory location
+here is "c". Reasoning only about this snippet, there is neither the data race
+in space (the race on "g") nor in time (the future race on "c"). Hence, the
+snippet will have sequentially consistent behaviour, and the value returned by
+"!c" will be "42".
+
+\subsection{s:mm_memory_safety}{Memory safety}
+
+The OCaml memory model guarantees that even for programs with data races,
+memory safety is preserved. While programs with data races may observe
+non-sequentially consistent behaviours, they will not crash. Let us look at an
+example of how to utilise LDRF to argue about memory safety in the presence of
+data races. Let us take a variant of the message passing example seen earlier
+(section \ref{s:mm_msg_passing}).
+
+\begin{caml_example*}{verbatim}
+let r = ref None
+let d1 () =
+  r := Some "Hello, world!"
+let d2 () =
+  match !r with
+  | None -> ()
+  | Some s -> print_endline s
+\end{caml_example*}
+
+The program uses a non-atomic reference "r" to share a message between domains.
+It is easy to see that there is a data race on "r". For the value "Some
+\"Hello, world!\"", OCaml allocates two objects -- one for the string and the
+other for the "Some" constructor. Both of these objects are first allocated
+uninitialised and then initialised with initialising writes. Are we guaranteed
+that, when "d2" sees the "Some" variant when reading "r", the objects have been
+fully initialised?
+
+The answer is yes, and we can use LDRF to reason about this. Let us refactor the
+program to make the newly allocated objects explicit.
+
+\begin{caml_example*}{verbatim}
+let r = ref None
+let d1 () =
+  let s = "Hello, world!" in
+  let o = Some s in
+  r := o
+let d2 () =
+  let o = !r in
+  match o with
+  | None -> ()
+  | Some s -> print_endline s
+\end{caml_example*}
+
+Observe that if "d2" does not witness the initialising writes on "s" or "o",
+the program may crash. LDRF allows us to reason about the program behaviour for
+a subset of memory locations. Let us ignore "r" and only consider "s" and "o".
+The objects "s" and "o" have initialising writes in "d1" and reads in "d2". The
+happens-before definition (section \ref{s:happens_before}) says that the
+initial write to a location precedes any other actions on that location. Hence,
+there is a happens-before relationship between the initial writes to "s" and
+"o" in "d1" and the reads in "d2". Hence, there is no data race on "s" and "o",
+and the program fragment that uses "s" and "o" will have sequentially
+consistent behaviour.
```

Comment:
Of course, this has to be rewritten in light of my comment above.

Reply:
I've removed this example as it relied on the reasoning through happens-before. The memory safety still holds but not due to happens-before reasoning.